### ISSUE TITLE
Rule request: No explicit `== true` or `== false`

### ISSUE THREAD

**Issue by mdiep** [realm/SwiftLint#1502]
I occasionally come across code like this:

```swift
if condition == true { … }
```

which just be written as:

```swift
if condition { … }
```

I'd love to have a rule that disallowed `== true` and `== false`. (And `!= true`, `!= false`, which I haven't seen, but would be far worse.)

**Comment by jpsim** [comment]
Agreed. This is a good starter rule to write if someone's interested in contributing to SwiftLint but finds other requested rules daunting.

Might want to also validate this rule with the boolean literal on the left hand side:

```swift
if true == condition {}
if false != condition {}
```

**Comment by SDGGiesbrecht** [comment]
> This is a good starter rule.

I’m not so sure. It’s more complicated than it looks at first glance.

For example, what about `Bool?`?

```swift
func someFunction() -> Bool? {
    return true
}

if someFunction() == true {
    // Should the rule trigger?
}

if someFunction() {
    // This will not compile.
}

if someFunction ?? false {
    // This would satisfy both the rule and the compiler,
    // but is this really what we want?
}
```

**Comment by masters3d** [comment]
I'd like to add to also throw a violation when comparing to the same object.

```swift
if someObj == someObj {} // always true
if objA != objA {} // always false
```

**Comment by jpsim** [comment]
Ah, cursed optionals strike again! @masters3d could you please file a new ticket for that, as its implementation could be entirely different than the original request.

**Comment by mohamede1945** [comment]
For the case where comparing with `Optional<Bool>` . I think it's clearer to use `?? true|false` operator than just compare with `==  true|false`.
If I look again at @SDGGiesbrecht 's example at the call site.
```swift
if someFunction() == true {
}
```
It's not clear for the reader that someFunction() might return `nil`.

But
```swift
if someFunction() ?? false {
}
```
It's very clear that the function might return `nil` and also clear what to do in that case.

**Comment by marcelofabri** [comment]
You could also use an `extension` (since Swift 3.1):

```swift
extension Optional where Wrapped == Bool {
    var isTrue: Bool {
        switch self {
        case .none:
            return false
        case .some(let value):
            return value
        }
    }
}
```

**Comment by mohamede1945** [comment]
But that extension assumes `nil` is equivalent to `false` which isn't.

**Comment by marcelofabri** [comment]
IMO it doesn't assume. It'd be the same as using `== true` or `?? false`. If you need another behavior based on if it's `nil`, it's just a matter of don't using the extension.

**Comment by koenpunt** [comment]
Ternary operations should also be considered;

```swift
index < activeIndex ? true : false
```
